src/d3a/models/myco_matcher/external_matcher.py

```diff
@@ -1,140 +1,179 @@
-                    # Offer or Bid either don't belong to market or were already matched
-                    raise InvalidBidOfferPair
-
-                if record.get("market_id") not in validated_records:
-                    validated_records[record.get("market_id")] = []
-
-                validated_records[record.get("market_id")].append(BidOfferMatch(
-                    bid,
-                    record.get("selected_energy"),
-                    offer,
-                    record.get("trade_rate")))
-            except InvalidBidOfferPair as ex:
-                # If validation fails or offer/bid were consumed
-                response_dict["status"] = "fail"
-                response_dict["message"] = "Validation Error"
-                logging.exception(f"Bid offer pair validation failed with error {ex}")
-                break
-        if response_dict["status"] == "success":
+        try:
+            validated_records = self._get_validated_bid_offer_match_list(recommendations)
             for market_id, records in validated_records.items():
                 market = self.markets_mapping.get(market_id)
-                if market.readonly:
-                    # The market has just finished
+                try:
+                    market.match_recommendation(records)
+                except (OfferNotFoundException, BidNotFoundException):
+                    # If the offer or bid have just been consumed
                     continue
-                market.match_recommendation(records)
-        self.myco_ext_conn.publish_json(channel, response_dict)
+        except Exception as ex:
+            response_data["status"] = "fail"
+            response_data["message"] = "Validation Error"
+            if not isinstance(ex, MycoValidationException):
+                logging.exception("Bid offer pair matching failed.")
+
+        self.myco_ext_conn.publish_json(channel, response_data)
 
-    def get_simulation_id(self, message):
-        """Publish the simulation id to the Myco client."""
+    def publish_simulation_id(self, message):
+        """Publish the simulation id to the redis myco client.
+
+        At the moment the id of the simulations run by the cli is set as ""
+        however, this function guarantees that the myco is aware of the running collaboration id
+        regardless of the value set in d3a.
+        """
 
         channel = "external-myco/get-simulation-id/response"
         self.myco_ext_conn.publish_json(channel, {"simulation_id": self.simulation_id})
 
     def publish_event_tick_myco(self):
         """Publish the tick event to the Myco client."""
 
-        channel = f"external-myco/{d3a.constants.COLLABORATION_ID}/response/events/"
         data = {"event": "tick"}
-        self.myco_ext_conn.publish_json(channel, data)
+        self.myco_ext_conn.publish_json(self.events_channel, data)
 
-    def publish_market_cycle_myco(self):
+    def publish_event_market_myco(self):
```

Comment:
What's a market event? Why has it been renamed from the previous "market_cycle"?

Reply:
I was trying to homogenize the handlers names, publish_tick_myco, publish_finish_myco...
In the myco and api-client, they are aware of the market event, not the market cycle, so I named it here like that
However I suppose that it should be called the same as we call the event in d3a, so I will restore it